In [24]:
import pickle
import numpy as np
from sklearn.model_selection import train_test_split

#data = pickle.load( open( "./template/trainingdata.p", "rb" ), encoding="latin1" )
#data = pickle.load(open("trainingdata_all.pickle", "rb"), encoding="latin1")
data = pickle.load(open("trainingdata_add2.pickle", "rb"))
n_images = len(data)
#test, training = data[0:int(n_images/3)], data[int(n_images/3):]
training, test = train_test_split(data, test_size=0.33, random_state=321)

def get_training_data():

    trX = np.array([np.reshape(a[2],a[2].shape[0]**2) for a in training])
    print(np.shape(trX)[1])
    trY = np.zeros((len(training)),dtype=np.float)
    for i, data in enumerate(training):
        trY[i] = float(data[0])
    return trX, trY

def get_test_data():
    teX = np.array([np.reshape(a[2],a[2].shape[0]**2) for a in test])
    teY = np.zeros((len(test)),dtype=np.float)
    for i, data in enumerate(test):
        teY[i] = float(data[0])
    return teX,teY


In [25]:
n_images

1131

In [26]:
from keras.models import Sequential
from keras.layers import Dense
#from sklearn.model_selection import train_test_split

import numpy as np
#import pandas as pd
import tensorflow as tf
#import pickle
#from get_image_data import *

class DNN_Driver():
    def __init__(self):
        self.trX = None
        self.trY = None
        self.teX = None
        self.teY = None
        self.model = None

    def tf_learn(self):
        self.trX, self.trY = get_training_data()
        self.teX, self.teY = get_test_data()

        seed = 0
        np.random.seed(seed)
        tf.random.set_seed(seed)

        self.model=Sequential()
        self.model.add(Dense(512, input_dim=np.shape(self.trX)[1], activation='relu'))
        self.model.add(Dense(64, activation='relu'))
        self.model.add(Dense(1))

        self.model.compile(loss='mean_squared_error', optimizer='adam')

        self.model.fit(self.trX, self.trY, epochs=6, batch_size=1)
        
        Y_prediction = self.model.predict(self.teX).flatten()

        for i in range(len(self.teY)):
            label = self.teY[i]
            pred = Y_prediction[i]
            print("label:{:.2f}, pred:{:.2f}".format(label, pred))
        return

    
    def get_direction(img):
        print(img.shape)
    #    img = np.array([np.reshape(img,img.shape**2)])
        ret =  self.model.predict(np.array([img]))
        return ret


        
        
    def predict_direction(self, img):
        print(img.shape)
#        img = np.array([np.reshape(img,img.shape**2)])
        ret =  self.model.predict(np.array([img]))
        return ret

    def get_test_img(self):
        img = self.teX[10]
        return img


In [27]:
print(len(training), len(test))

757 374


In [28]:
dnn_driver = DNN_Driver()
dnn_driver.tf_learn()

256
Epoch 1/6
757/757 [==============================] - 1s 2ms/step - loss: 666.0933
Epoch 2/6
757/757 [==============================] - 1s 1ms/step - loss: 1.5435
Epoch 3/6
757/757 [==============================] - 1s 1ms/step - loss: 0.1681
Epoch 4/6
757/757 [==============================] - 1s 1ms/step - loss: 0.1258
Epoch 5/6
757/757 [==============================] - 1s 1ms/step - loss: 0.0964
Epoch 6/6
757/757 [==============================] - 1s 1ms/step - loss: 0.0687
label:0.00, pred:0.12
label:1.00, pred:1.26
label:0.00, pred:-0.18
label:0.00, pred:0.31
label:-1.00, pred:-0.62
label:0.50, pred:1.23
label:-0.50, pred:-0.43
label:0.50, pred:0.74
label:0.00, pred:0.62
label:0.00, pred:0.12
label:0.00, pred:0.17
label:0.00, pred:-0.01
label:0.00, pred:0.09
label:-0.50, pred:-0.28
label:0.00, pred:0.00
label:-0.50, pred:-0.58
label:0.00, pred:0.02
label:0.50, pred:0.68
label:-1.00, pred:-0.78
label:1.00, pred:1.09
label:0.00, pred:0.01
label:0.00, pred:-0.03
label:0.50, pred:

## 기록 (add2)
- trainingdata_add2.pickle, test_size = 0.33, random_state = 123, epoch 6
    - Epoch 1/6
    - 757/757 [==============================] - 1s 2ms/step - loss: 472.0614
    - Epoch 2/6
    - 757/757 [==============================] - 1s 2ms/step - loss: 3.4715
    - Epoch 3/6
    - 757/757 [==============================] - 1s 2ms/step - loss: 0.4602
    - Epoch 4/6
    - 757/757 [==============================] - 1s 1ms/step - loss: 0.2229
    - Epoch 5/6
    - 757/757 [==============================] - 1s 1ms/step - loss: 0.1586
    - Epoch 6/6
    - 757/757 [==============================] - 1s 1ms/step - loss: 0.1518
    
- trainingdata_add2.pickle, test_size = 0.33, random_state = 321, epoch 6
    - Epoch 1/6
    - 757/757 [==============================] - 1s 1ms/step - loss: 666.0933
    - Epoch 2/6
    - 757/757 [==============================] - 1s 1ms/step - loss: 1.5435
    - Epoch 3/6
    - 757/757 [==============================] - 1s 1ms/step - loss: 0.1681
    - Epoch 4/6
    - 757/757 [==============================] - 1s 1ms/step - loss: 0.1258A: 
    - Epoch 5/6
    - 757/757 [==============================] - 1s 2ms/step - loss: 0.0964
    - Epoch 6/6
    - 757/757 [==============================] - 1s 1ms/step - loss: 0.0687

## 기록 (add, 괜찮았던 조합)
- trainingdata_all.pickle, test_size = 0.33, random_state = 123, epoch = 2 <0.02>
    - Epoch 1/2
    - 4920/4920 [==============================] - 7s 1ms/step - loss: 38.6281
    - Epoch 2/2
    - 4920/4920 [==============================] - 7s 1ms/step - loss: 0.1972
- trainingdata_all.pickle, test_size = 0.33, random_state = 321, epoch = 2 <0.03>
    - Epoch 1/2
    - 4920/4920 [==============================] - 7s 1ms/step - loss: 68.4014
    - Epoch 2/2
    - 4920/4920 [==============================] - 7s 1ms/step - loss: 0.1841
- trainingdata_all.pickle, test_size = 0.35, random_state = 321, epoch = 2 <0.02> <비추천?>
    - Epoch 1/2
    - 4773/4773 [==============================] - 7s 1ms/step - loss: 57.7040
    - Epoch 2/2
    - 4773/4773 [==============================] - 7s 1ms/step - loss: 0.2754
- trainingdata_all.pickle, test_size = 0.33, random_state = 747, epoch = 2 <0.03>
    - Epoch 1/2
    - 4920/4920 [==============================] - 7s 1ms/step - loss: 84.2031
    - Epoch 2/2
    - 4920/4920 [==============================] - 7s 1ms/step - loss: 0.1877
- trainingdata_add.pickle, test_size = 0.33, random_state = 123, epoch = 6
    - Epoch 1/6
    - 753/753 [==============================] - 1s 1ms/step - loss: 585.2233
    - Epoch 2/6
    - 753/753 [==============================] - 1s 1ms/step - loss: 1.2138
    - Epoch 3/6
    - 753/753 [==============================] - 1s 1ms/step - loss: 0.3951
    - Epoch 4/6
    - 753/753 [==============================] - 1s 1ms/step - loss: 0.2933
    -Epoch 5/6
    - 753/753 [==============================] - 1s 1ms/step - loss: 0.1878
    - Epoch 6/6
    - 753/753 [==============================] - 1s 1ms/step - loss: 0.1411